# Exercise 05. (Take-home) Advanced Topic: Adding an in-game Recommender using RLlib

© 2019-2022, Anyscale. All Rights Reserved <br>
📖 [Back to Table of Contents](./ex_00_rllib_notebooks_table_of_contents.ipynb)<br>
➡️ [Next notebook](./ex_06_rllib_end_to_end_demo.ipynb) <br>
⬅️ [Previous notebook](./ex_04_offline_rl_with_rllib.ipynb) <br>

### Learning objectives
In this this tutorial, you will learn how to:

 * [Intro RecSys with RL](#recsys_rl)
 * [Create a RecSys RL environment](#recsys_env)
 * [Train a Contextual Bandit on the environment](#cb)
 * [Train using a RL Offline algorithm on the environment](#offline)
 
 
 find RLlib algos to train policy models on environments.
 

## Intro RecSys with RL <a class="anchor" id="recsys_rl"></a>

A Recommender System <b>(RecSys)</b> suggests items that are most pertinent to a particular user.  Examples of recommender systems include:
<ul>
    <li>Movie/video/music recommendations (Netflix/YouTube/Spotify)</li>
    <li>Online shopping recommendations (Amazon/Shopify)</li>
    <li>Filtering your feed as you scroll (Twitter/Instagram)</li>
</ul>

<b>Two main approaches to training algorithms</b> for RecSys are: 
<ol>
    <li>Traditional Machine Learning <b>(ML)</b></li>
    <li>Reinforcement Learning <b>(RL)</b></li>
    </ol>

<b>In the traditional ML method</b>, data is gathered about users interactions with items.  Matrix factorization is often used to turn this data into trainable representations of users and items ("embeddings" or features or X's), and the views or actions by users of those items (dependent variable or y's). A ranking algorithm (e.g. collaborative filtering) is trained on all the data at once as if all the actions happend during the same time step.  Such a <b><i>static</i> </b>, fixed model is useful when there are millions of items and users, since learning from all data at once in <b>batch learning</b> is efficient.

<b>In RL</b>, users interact with offers repeatedly over time.  This <b><i>dynamic</i></b> model is iteratively trained based on 
the last observation, action, and reward.  One caveat with RL, since a recommendation needs to be calculated at every time step, only a pre-selected handful of top candidate items per user (from the traditional ML ranking model) is presented in the simulation environment.  

Both Online RL and Offline RL are important parts of modern Recommender Systems. 
<b>Online RL in RecSys</b> takes as input the top-K ranked recommendations per user, keeps the user and item feature embeddings as context, but optimizes a sequence of user in-session interactions <b>real-time</b> using live results.  

<img src="./images/recsys_overview.png" width="100%" />

<b>Offline RL in RecSys</b> typically uses log files to explore the last recommender model put into production, implicitly through the data.  “Serendipitous” aspects of user experience can be explored through offline RL, since random actions the user did not historically take can be tried in the simulation.

<div class="alert alert-block alert-success">    
    <b>💡 Online vs Offline RL, when the algorithm learning from an environment is: </b> <br><br>
    ✔️ live (typically gaming platforms or complex systems simulations), this is called <b>online RL</b> and evaluation during training is <b>on-policy</b>. <br><br>
    ✔️ trained using log files (users interactions with items), this is called <b>offline RL</b> and evaluation during training is <b>off-policy</b>, because the policy (RL word for model) used to log the data is different from the policy used to explore the data. </b> 
</div>

<br>

## Create a RecSys RL Environment <a class="anchor" id="recsys_env"></a>

As we learned in the first 2 lessons, the first step to training a RL RecSys policy model is to create an <b>environment</b> that can interact with a RL Algorithm to train a recommender agent. 

In this notebook, we will use <b><a href="https://github.com/google-research/recsim">Google's RecSim environment</a></b>, which was developed for the YouTube recommendation problem.  The environment is <i>Timelimit-based</i>, meaning the termination condition for an episode will be after a fixed number (60) of videos are watched. The RecSim environment consists of:

<img src="./images/recsim_environment.png" width="90%" />

* <b>Document Model</b>, in the range [0, 1].  
<ul>
    <li>On the 0-end of the scale, <b>"sweet"</b> documents lead to large amounts of <b>"click bait"</b> or immediate engagement. Sweetness values are drawn from ln Normal(μsweet, σsweet).</li>
    <li>On the 1-end of the scale, documents termed <b>kale</b>, are less click-bait, but tend to <b>increase user long-term satisfaction</b>. Kale values are drawn from ln Normal(μkale, σkale)</li>
    <li>Mixed doc values are drawn from linear interpolation between parameters of the two distributions in proportion to their kaleness.</li>
    </ul>
* <b>User Model</b>, empty vectors.  <i>Since user features are so business-dependent, RecSim out-of-the-box does not include any user features.</i>  User features, typically imported from real data, could contain: 
<ul>
    <li>evolving, unknown contexts (interests, preferences, satisfaction, activity, mood)</li>
    <li>unobservable events that could impact user behavior (personalized promotions, interuptions that cause turning off a video such as because someone rang their doorbell).</li>
    </ul>
* <b>Rewards</b>, or user satisfaction after their choice, modeled in the range [0, 1] that stochastically (and slowly) increases or decreases with the consumption of different types of content; kale or sweetness.  


<b>RLlib comes with 3 RecSim environments</b>  <br>
<div class="alert alert-block alert-success">    
👉 <b>Long Term Satisfaction</b> (used in this tutorial) <br>
- Interest Evolution <br>
- Interest Exploration <br>
</div>

<br>

In [1]:
# import libraries
import numpy as np
from scipy.stats import linregress, sem
import ray
from ray import tune
from ray.tune.logger import pretty_print
print(f"ray: {ray.__version__}")

# silence the many tensorflow warnings
import logging, os
logging.disable(logging.WARNING)
os.environ["TF_CPP_MIN_LOG_LEVEL"] = "3"
import tensorflow as tf
import gym
import recsim

print(f"tensorflow: {tf.__version__}")
print(f"gym: {gym.__version__}")

# Import the built-in RecSim exapmle environment: "Long Term Satisfaction", ready to be trained by RLlib.
from ray.rllib.examples.env.recommender_system_envs_with_recsim import LongTermSatisfactionRecSimEnv


ray: 3.0.0.dev0
tensorflow: 2.6.0
gym: 0.21.0


In [2]:
# Create a RecSim instance using the following config parameters 
lts_10_1_env = LongTermSatisfactionRecSimEnv({
    "num_candidates": 10,  # The number of possible documents/videos/candidates that we can recommend
    "slate_size": 1, # The number of recommendations that we will be making
    # Set to False for re-using the same candidate doecuments each timestep.
    "resample_documents": False,
    # Convert MultiDiscrete actions to Discrete (flatten action space).
    # e.g. slate_size=2 and num_candidates=10 -> MultiDiscrete([10, 10]) -> Discrete(100)  # 10x10
    "convert_to_discrete_action_space": True,
})

# inspect the gym spec for the environment
print(f"env: {lts_10_1_env}")

# print gym Spaces
if isinstance(lts_10_1_env.action_space, gym.spaces.Space):
    print()
    print(f"action space: {lts_10_1_env.action_space}")
if isinstance(lts_10_1_env.observation_space, gym.spaces.Space):
    print()
    print(f"observation space: {lts_10_1_env.observation_space}") 
        
# Note: the action space is discrete with 10 possible actions.
# Note: observations consist of 1 user, 10 docs


env: <_RecSimEnv<MultiDiscreteToDiscreteActionWrapper<RecSimObservationSpaceWrapper<RecSimResetWrapper<RecSimGymEnv instance>>>>>

action space: Discrete(10)

observation space: Dict(user:Box([], [], (0,), float32), doc:Dict(0:Box([0.], [1.], (1,), float32), 1:Box([0.], [1.], (1,), float32), 2:Box([0.], [1.], (1,), float32), 3:Box([0.], [1.], (1,), float32), 4:Box([0.], [1.], (1,), float32), 5:Box([0.], [1.], (1,), float32), 6:Box([0.], [1.], (1,), float32), 7:Box([0.], [1.], (1,), float32), 8:Box([0.], [1.], (1,), float32), 9:Box([0.], [1.], (1,), float32)), response:Tuple(Dict(click:Discrete(2), engagement:Box(0.0, 100.0, (), float32))))


In [3]:
# Modifying wrapper around the LTS (Long Term Satisfaction) env:
# - allows us to tweak the user model (and thus: reward behavior)
# - adds user's current satisfaction value to observation

class LTSWithStrongerDissatisfactionEffect(gym.ObservationWrapper):

    def __init__(self, env):
        # Tweak incoming environment.
        env.environment._user_model._user_sampler._state_parameters.update({
            "sensitivity": 0.058,
            "time_budget": 120,
            "choc_stddev": 0.1,
            "kale_stddev": 0.1,
            #"innovation_stddev": 0.01,
            #"choc_mean": 1.25,
            #"kale_mean": 1.0,
            #"memory_discount": 0.9,
        })

        super().__init__(env)

        # Adjust observation space.
        if "response" in self.observation_space.spaces:
            self.observation_space.spaces["user"] = gym.spaces.Box(0.0, 1.0, (1, ), dtype=np.float32)
            for r in self.observation_space["response"]:
                if "engagement" in r.spaces:
                    r.spaces["watch_time"] = r.spaces["engagement"]
                    del r.spaces["engagement"]
                    break

    def observation(self, observation):
        if "response" in self.observation_space.spaces:
            observation["user"] = np.array([self.env.environment._user_model._user_state.satisfaction])
            for r in observation["response"]:
                if "engagement" in r:
                    r["watch_time"] = r["engagement"]
                    del r["engagement"]
        return observation


# Add the wrapping around 
tune.register_env("modified-lts", 
                  lambda env_config: LTSWithStrongerDissatisfactionEffect(
                      LongTermSatisfactionRecSimEnv(env_config)))

print("ok; registered the string 'modified-lts' to be used in RLlib configs (see below)")


ok; registered the string 'modified-lts' to be used in RLlib configs (see below)


In [4]:
# How to make RecSim environments?

# 20 candidates (instead of 10) and a slate-size of 2.
# Try to strengthen the dissatisfaction effect on the engagement

# 1. define a config dictionary
env_config_20 = \
    {
        # The number of possible documents/videos/candidates that we can recommend
        "num_candidates": 20,  
        # The number of recommendations that we will be making
        "slate_size": 2,  # MultiDiscrete([20, 20]) -> Discrete(400)
        # Set to False for re-using the same candidate doecuments each timestep.
        "resample_documents": True,
        # Convert MultiDiscrete actions to Discrete (flatten action space).
        "convert_to_discrete_action_space": True,
        # Wrap observations for RLlib bandit: Only changes dict keys ("item" instead of "doc").
        "wrap_for_bandits": True,}

# 2. create a RecSim environment
lts_20_2_env = LongTermSatisfactionRecSimEnv(env_config_20)

# 3. create a modified RecSim environment
# LTSWithStrongerDissatisfactionEffect(recsim_env)
modified_lts_20_2_env = \
    LTSWithStrongerDissatisfactionEffect(lts_20_2_env)

print(type(modified_lts_20_2_env))


<class '__main__.LTSWithStrongerDissatisfactionEffect'>


<b> Check intuition about environment</b>

In the cell below, we run for 1K episodes and record the reward slopes for 3 scenarios:
    <ul>
        <li>Random policy: pick randomly every time</li>
        <li>Greedy sweet policy: always pick the sweetest item </li>
        <li>Greedy kale policy: always pick the kaleist item </li>
    </ul>

The result below shows random is worst; greedy sweet is next worst with negative reward slope; and greedy kale is best with very small positive reward slope.

In [5]:
# Check slopes with more complicated env
lts_20_2_env = \
    LTSWithStrongerDissatisfactionEffect(
        LongTermSatisfactionRecSimEnv(env_config_20))

# Capture slopes of all trendlines over all episodes.
slopes = []

# Run 1000 episodes
num_episodes = 1000
for _ in range(num_episodes):
    obs = lts_20_2_env.reset()  # Reset environment to get initial observation:

    # Compute actions that pick doc with highest/lowest feature value.
    action_sweetest = np.argmax(obs['item'])
    action_kaleiest = np.argmin(obs['item'])

    # Play one episode.
    done = False
    rewards = []
    while not done:
        action = lts_20_2_env.action_space.sample()
        # action = action_sweetest
        # action = action_kaleiest

        obs, reward, done, _ = lts_20_2_env.step(action)
        rewards.append(reward)

    # Create linear model of rewards over time.
    reward_linreg = linregress(np.array((range(len(rewards)))), np.array(rewards))
    slopes.append(reward_linreg.slope)

print(f"after {num_episodes} episodes, reward slope: {np.mean(slopes)}")

# random slope: 2.0507302101210928e-05          # smallest positive
# greedy sweet slope: -0.00022956105559631047   # negative
# greedy kale slope:  8.599181822029186e-05     # larger positive

after 1000 episodes, reward slope: -8.737139146441298e-05


<b>Get an environment baseline</b>

Above we got slopes, to help with intution.  But, it is always best practice, before training an algorithm, to run through the environment and record the mean reward as a baseline.

<div class="alert alert-block alert-success">
💡 For this environment, we will calculate 3 baselines: <br>
    <ul>
        <li>Random baseline: pick randomly every time</li>
        <li>Greedy sweet baseline: always pick the sweetest item </li>
        <li>Greedy kale baseline: always pick the kaleist item </li>
    </ul>
</div>

In [6]:
# Function that measures and outputs the random baseline reward.
# This is the expected accumulated reward per episode, if we act randomly (recommend random items) at each time step.
def measure_random_performance_for_env(env, 
                                       baseline_type="random",
                                       episodes=1000, verbose=False):

    # Reset the env.
    obs = env.reset()

    # Number of episodes already done.
    num_episodes = 0
    # Current episode's accumulated reward.
    episode_reward = 0.0
    # Collect all episode rewards here to be able to calculate a random baseline reward.
    episode_rewards = []

    # Enter while loop (to step through the episode).
    while num_episodes < episodes:
        # Produce an action
        action_random = env.action_space.sample()
        action_sweetest = np.argmax(obs['item'])
        action_kaleiest = np.argmin(obs['item'])
        if baseline_type == "random":
            action = action_random
        elif baseline_type == "sweetest":
            action = action_sweetest
        elif baseline_type == "kaleist":
            action = action_kaleiest

        # Send the action to the env's `step()` method to receive: obs, reward, done, and info.
        obs, reward, done, _ = env.step(action)
        episode_reward += reward

        # Check, whether the episde is done, if yes, reset and increase episode counter.
        if done:
            if verbose:
                print(f"Episode done - accumulated reward={episode_reward}")
            elif num_episodes % 99 == 0:
                print(f" {num_episodes} ", end="")
            elif num_episodes % 9 == 0:
                print(".", end="")
            num_episodes += 1
            env.reset()
            episode_rewards.append(episode_reward)
            episode_reward = 0.0

    # Print out and return mean episode reward (and standard error of the mean).
    env_mean_random_reward = np.mean(episode_rewards)
    env_sd_reward = sem(episode_rewards)

    print(f"\n\nMean {baseline_type} baseline reward: {env_mean_random_reward:.2f}+/-{sem(episode_rewards):.2f}")

    return env_mean_random_reward, sem(episode_rewards)


In [7]:
# Calculate a baseline type: random, sweetest, or kaleist
lts_20_2_env_mean_random_reward, _ = \
    measure_random_performance_for_env(lts_20_2_env, 
                                       baseline_type="sweetest",
                                       episodes=100)

# Random baseline: 1158.00+/-1.21    # next biggest
# Sweetest baseline: 1164.24+/-1.18  # biggest
# Kaleist baseline: 1085.04+/-1.05   # smallest

 0 .......... 99 

Mean sweetest baseline reward: 1162.53+/-1.03


## Train a Contextual Bandit on the environment <a class="anchor" id="cb"></a>

A Bandit session is one where we have an opportunity to recommend the user an item and observe their behaviour. We receive a reward if they click.

<ol>
    <li>Open RLlib docs <a href="https://docs.ray.io/en/master/rllib/rllib-algorithms.html">and navigate to the Algorithms page.</a></li>
    <li>Scroll down and click url of algo you want to use, e.g. <i><b>Bandits</b></i></li>
    <li>On the <a href="https://docs.ray.io/en/master/rllib/rllib-algorithms.html#bandits">algo docs page </a>, click on the link <i><b>Implementation</b></i>.  This will open the <a href="https://github.com/ray-project/ray/blob/master/rllib/algorithms/bandit/bandit.py">algo code file on github</a>.</li>
    <li>Search the github code file for the word <i><b>config</b></i></li>
    <li>Typically the docstring example will show: </li>
    <ol>
        <li>Example code implementing RLlib API, then </li>
        <li>Example code implementing Ray Tune API.</li>
    </ol>
    <li>Scroll down to the config <b>__init()__</b> method</li>
    <ol>
            <li>Algorithm default hyperparameter values are here.</li>
    </ol>
    </ol>

In [8]:
# Select RLlib Bandit algorithm w/Upper Confidence Bound (UCB) exploration
# and find that algorithm's config class

# config is an object instead of a dictionary since Ray version >= 1.13
from ray.rllib.algorithms.bandit import BanditLinUCBConfig

In [9]:
# Choose your config settings and instantiate a config object with those settings
# Define algorithm config values
env_name = "modified-lts"
evaluation_interval = 2   #100, num training episodes to run between eval steps
evaluation_duration = 2  #100, num eval episodes to run for the eval step
num_workers = 4          # +1 for head node, num parallel workers or actors for rollouts
num_gpus = 0             # num gpus to use in the cluster
num_envs_per_worker = 1  #1, no vectorization of environments to run at same time

# Define trainer runtime config values
checkpoint_freq = evaluation_interval # freq save checkpoints >= evaulation_interval
checkpoint_at_end = True                # always save last checkpoint
relative_checkpoint_dir = "my_LinUCB_logs" # redirect logs instead of ~/ray_results/
random_seed = 415
# Set the log level to DEBUG, INFO, WARN, or ERROR 
log_level = "ERROR"


In [10]:
# # uncomment below to see the long list of default config values
# print(f"Bandit's default config is:")
# print(pretty_print(BanditLinUCBConfig().to_dict()))

# Create a defaut config:
bandit_config = BanditLinUCBConfig()

# Set it up for the correct environment:
bandit_config.environment(env=env_name, env_config=env_config_20)

# Decide if you want torch or tensorflow DL framework
bandit_config.framework("tf2", eager_tracing=True)

# Set up evaluation as follows:
bandit_config.evaluation(
    # Run evaluation once per `train()` call.
    evaluation_interval=evaluation_interval,
    # Use separate resources (RLlib rollout workers).
    evaluation_num_workers=num_workers,
    # Run 20 episodes per evaluation (per iteration) -> 10 per eval worker (we have 2 eval workers).
    evaluation_duration=evaluation_duration,
    evaluation_duration_unit="timesteps",
    # # Run evaluation alternatingly with training (not in parallel).
    # evaluation_parallel_to_training=False,
)
# Create a new training config
# override certain default algorithm config values
bandit_config\
    .rollouts(num_rollout_workers=num_workers, num_envs_per_worker=num_envs_per_worker)\
    .resources(num_gpus=num_gpus, )\
    .debugging(seed=random_seed, log_level=log_level)

print(type(bandit_config))


<class 'ray.rllib.algorithms.bandit.bandit.BanditLinUCBConfig'>


In [11]:
# TODO - why is ray tune for bandit so slow?

# # Example using Ray tune API (`tune.run()`) until some stopping condition is met.
# # This will create one (or more) Algorithms under the hood automatically w/o us having to
# # build these algos from the config.

# experiment_results = tune.run(
#     "BanditLinUCB",

#     # training config params (translated into a python dict!)
#     config=bandit_config.to_dict(),

#     # Stopping criteria whichever occurs first: average reward over training episodes, or ...
#     stop={
#         "training_iteration": 10,     # stop after n training iterations (calls to `Algorithm.train()`)
#         #"episode_reward_mean": 400, # stop if average (sum of) rewards in an episode is 400 or more
#         #"timesteps_total": 100000,  # stop if reached 100,000 sampling timesteps
#     },  

#     # redirect logs instead of default ~/ray_results/
#     local_dir=relative_checkpoint_dir,
         
#     # Every how many train() calls do we create a checkpoint?
#     checkpoint_freq=checkpoint_freq,
#     # Always save last checkpoint (no matter the frequency).
#     checkpoint_at_end=checkpoint_at_end,

#     ###############
#     # Note about Ray Tune verbosity.
#     # Screen verbosity in Ray Tune is defined as verbose = 0, 1, 2, or 3, where:
#     # 0 = silent
#     # 1 = only status updates, no logging messages
#     # 2 = status and brief trial results, includes logging messages
#     # 3 = status and detailed trial results, includes logging messages
#     # Defaults to 3.
#     ###############
#     verbose=3,
                   
#     # Define what we are comparing for, when we search for the
#     # "best" checkpoint at the end.
#     metric="episode_reward_mean",
#     mode="max",
# )

# print("Training completed.")
# print("Best checkpoint: ", experiment_results.best_checkpoint)


In [12]:
# To start fresh, restart Ray in case it is already running
if ray.is_initialized():
    ray.shutdown()
    
# Use the config object's `build()` method for generating
# an RLlib Algorithm instance that we can then train.
linucb_algo = bandit_config.build()
print(f"Algorithm type: {type(linucb_algo)}")

# train the Bandit Algorithm instance for n iterations (timesteps)
rewards = []
n_timesteps = 100
for i in range(n_timesteps):
    # Run a single timestep in the environment and update
    # the model immediately on the received reward.
    result = linucb_algo.train()
    
    # Extract reward from results.
    rewards.append(result["episode_reward_mean"])
    
    if i % 99 == 0:
        print(f"\nIteration={i}, ", end="")
        print(f"Mean Bandit Reward={result['episode_reward_mean']:.2f}",end="")
        try:
            print(f"+/-{np.std(rewards[4:]):.2f}\n")
        except:
            print()
    elif num_episodes % 9 == 0:
        print(".", end="")

# To stop the Algorithm and release its blocked resources, use:
linucb_algo.stop()


2022-08-05 08:16:41,512	ERROR services.py:1376 -- Failed to start the dashboard: Failed to start the dashboard, return code 1
 The last 10 lines of /tmp/ray/session_2022-08-05_08-16-39_599973_26052/logs/dashboard.log:
  File "/Users/christy/miniforge3/envs/rllib2/lib/python3.8/site-packages/aiohttp/web_urldispatcher.py", line 1200, in add_routes
    registered_routes.extend(route_def.register(self))
  File "/Users/christy/miniforge3/envs/rllib2/lib/python3.8/site-packages/aiohttp/web_routedef.py", line 98, in register
    resource = router.add_static(self.prefix, self.path, **self.kwargs)
  File "/Users/christy/miniforge3/envs/rllib2/lib/python3.8/site-packages/aiohttp/web_urldispatcher.py", line 1124, in add_static
    resource = StaticResource(
  File "/Users/christy/miniforge3/envs/rllib2/lib/python3.8/site-packages/aiohttp/web_urldispatcher.py", line 559, in __init__
    raise ValueError(f"No directory exists at '{directory}'") from error
ValueError: No directory exists at '/Users/

(RolloutWorker pid=26101) Metal device set to: Apple M1
(RolloutWorker pid=26099) Metal device set to: Apple M1
(RolloutWorker pid=26100) Metal device set to: Apple M1
(RolloutWorker pid=26098) Metal device set to: Apple M1
Metal device set to: Apple M1
(RolloutWorker pid=26122) Metal device set to: Apple M1
(RolloutWorker pid=26125) Metal device set to: Apple M1
(RolloutWorker pid=26123) Metal device set to: Apple M1
(RolloutWorker pid=26124) Metal device set to: Apple M1
Algorithm type: <class 'ray.rllib.algorithms.bandit.bandit.BanditLinUCB'>

Iteration=0, Mean Bandit Reward=nan+/-nan



/Users/christy/miniforge3/envs/rllib2/lib/python3.8/site-packages/numpy/core/_methods.py:233: RuntimeWarning: Degrees of freedom <= 0 for slice
  ret = _var(a, axis=axis, dtype=dtype, out=out, ddof=ddof,
/Users/christy/miniforge3/envs/rllib2/lib/python3.8/site-packages/numpy/core/_methods.py:194: RuntimeWarning: invalid value encountered in true_divide
  arrmean = um.true_divide(
/Users/christy/miniforge3/envs/rllib2/lib/python3.8/site-packages/numpy/core/_methods.py:226: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)



Iteration=99, Mean Bandit Reward=1164.80+/-1.48



In [ ]:
# Plot per-timestep (episode) rewards.
plt.figure(figsize=(10,7))
start_at = 0
smoothing_win = 200
x = list(range(start_at, len(rewards)))
y = [np.nanmean(rewards[max(i - smoothing_win, 0):i + 1]) for i in range(start_at, len(rewards))]
plt.plot(x, y)
plt.title("Mean reward")
plt.xlabel("Time/Training steps")

# Add mean random baseline reward (red line).
plt.axhline(y=lts_20_2_env_mean_random_reward, color="r", linestyle="-")

plt.show();


Comparing our algorithms over 100 timesteps:
<ul>
    <li>Kaleist baseline: 1084.49+/-1.04 </li>
    <li>Random baseline: 1158.00+/-1.21 </li>
    <b><li>Sweetest baseline: 1164.24+/-1.18 </li>
        <li>Bandit mean reward = 1164.80+/-1.48</li></b>
    </ul>


<div class="alert alert-block alert-success">
🤔 Notice that the bandit mean reward is suspiciously the same as the sweetest baseline! <br>
</div>

In this dummy Recsim environment, we did not have any user features.  This makes the contextual bandit without any user context, i.e. without any state.  A stateless bandit cannot remember things between timesteps, so it will be exactly the most greedy policy.  

So, it makes sense that the Bandit reward = Sweetest baseline!  The context-less bandit was unable to learn anything!


## Train using a RL Offline algorithm on the environment <a class="anchor" id="offline"></a>

CRR

In [13]:
# Import the config class of the algorithm, we would like to train with: CRR.
from ray.rllib.algorithms.crr import CRRConfig

In [13]:
# Create a defaut CRR config:
config = CRRConfig()

# Set it up for the correct environment:
# NOTE: We said above that we wouldn't really have an environment available (so how can
# we set one up here??).
# The following is only to tell the algorithm, which environment our offline data was actually taken from.
config.environment(env="modified-lts")
# If you really really don't have an environment, set `env=None` here and additionally define your action- and
# observation spaces.
# config.environment(env=None, action_space=..., observation_space=...)

#################################################
# This is the most important piece of code 
# in this notebook:
# It explains how to point your 
# algorithm to the correct offline data file
# (instead of a live-environment).
#################################################
config.offline_data(
    input_="dataset",
    input_config={
        # If you feel daring here, use the `pendulum_beginner.json` file instead of the expert one here.
        # You may need to train a little longer, then, in order to get a decent policy.
        # But since you have the actual Pendulum environment available for evaluation, you should be able
        # to perfectly stop learning once a good episode reward (> -300.0) has been reached.
        "paths": os.path.join(os.getcwd(), "offline_rl_data/pendulum_expert.json"),
        "format": "json",
    },
    # The (continuous) actions in our input files are already normalized
    # (meaning between -1.0 and 1.0) -> We don't have to do anything with them prior to
    # computing losses.
    actions_in_input_normalized=True,
)

# RLlib's CRR is a very new algorithm (since 1.13) and only supports
# the PyTorch framework thus far. We'll provide a tf version in the near future.
config.framework("torch")

# Set up evaluation as follows:
config.evaluation(
    # Run evaluation once per `train()` call.
    evaluation_interval=1,
    # Use a separate resource ("RLlib rollout worker")
    evaluation_num_workers=1,

    # # Use separate resources (RLlib rollout workers).
    # evaluation_num_workers=2,

    # Run 20 episodes per evaluation (per iteration) -> 10 per eval worker (we have 2 eval workers).
    evaluation_duration=20,
    evaluation_duration_unit="episodes",

    # Use a slightly different config for the evaluation:
    evaluation_config={
        # - Use a real environment (so we can fully trust the evaluation results, rewards, etc..)
        "input": "sampler",
        # - Switch off exploration for better (less stochastic) action computations.
        "explore": False,
    },

    # Run evaluation alternatingly with training (not in parallel).
    evaluation_parallel_to_training=False,
)

print(type(config))


<class 'ray.rllib.algorithms.crr.crr.CRRConfig'>


In [14]:
evaluation_interval = 100   #100, num training episodes to run between eval steps
verbosity = 2 # Tune screen verbosity

experiment_results = tune.run("CRR", 
                    
    # Stopping criteria whichever occurs first: average reward over training episodes, or ...
    stop={#"episode_reward_mean": 400, # stop if achieve 400 out of max 500
          "training_iteration": 100,  # stop if achieved 100 episodes
          # "timesteps_total": 100000,  # stop if achieved 100,000 timesteps
          },  
              
    # training config params
    config = config.to_dict(),
                    
    #redirect logs instead of default ~/ray_results/
    local_dir = relative_checkpoint_dir, #relative path
         
    # set frequency saving checkpoints >= evaulation_interval
    checkpoint_freq = checkpoint_freq,
    checkpoint_at_end=True,
         
    # Reduce logging messages
    verbose = verbosity,
    )

print("Training completed.")

(RolloutWorker pid=26101) 
(RolloutWorker pid=26101) systemMemory: 16.00 GB
(RolloutWorker pid=26101) maxCacheSize: 5.33 GB
(RolloutWorker pid=26101) 
(RolloutWorker pid=26100) 
(RolloutWorker pid=26100) systemMemory: 16.00 GB
(RolloutWorker pid=26100) maxCacheSize: 5.33 GB
(RolloutWorker pid=26100) 
(RolloutWorker pid=26098) 
(RolloutWorker pid=26098) systemMemory: 16.00 GB
(RolloutWorker pid=26098) maxCacheSize: 5.33 GB
(RolloutWorker pid=26098) 


NameError: name 'config' is not defined

(RolloutWorker pid=26122) 
(RolloutWorker pid=26122) systemMemory: 16.00 GB
(RolloutWorker pid=26122) maxCacheSize: 5.33 GB
(RolloutWorker pid=26122) 
(RolloutWorker pid=26125) 
(RolloutWorker pid=26125) systemMemory: 16.00 GB
(RolloutWorker pid=26125) maxCacheSize: 5.33 GB
(RolloutWorker pid=26125) 
(RolloutWorker pid=26123) 
(RolloutWorker pid=26123) systemMemory: 16.00 GB
(RolloutWorker pid=26123) maxCacheSize: 5.33 GB
(RolloutWorker pid=26123) 
(RolloutWorker pid=26124) 
(RolloutWorker pid=26124) systemMemory: 16.00 GB
(RolloutWorker pid=26124) maxCacheSize: 5.33 GB
(RolloutWorker pid=26124) 
(RolloutWorker pid=26099) 
(RolloutWorker pid=26099) systemMemory: 16.00 GB
(RolloutWorker pid=26099) maxCacheSize: 5.33 GB
(RolloutWorker pid=26099) 


### Exercises

1. 

### Homework

1. 

### References

* 

📖 [Back to Table of Contents](./ex_00_rllib_notebooks_table_of_contents.ipynb)<br>
➡️ [Next notebook](./ex_06_rllib_end_to_end_demo.ipynb) <br>
⬅️ [Previous notebook](./ex_04_offline_rl_with_rllib.ipynb) <br>